### PR TITLE
Reland "[ios][platform_view] Use CAShapeLayer as the mask to avoid software rendering #53072"

### PR DESCRIPTION
**Pull request by hellohuanlin**
The previous PR was reverted because it was on top of an outdated commit prior to ARC. This PR rebases the main. The only change is removing the `[super dealloc]` call that caused the compiler error. 

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/142813
Fixes https://github.com/flutter/flutter/issues/142830

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.mm [hunk 1]
```diff
@@ -252,15 +252,30 @@ - (instancetype)initWithFrame:(CGRect)frame screenScale:(CGFloat)screenScale {
...
 - (void)reset {
-  paths_.clear();
+  CGPathRelease(pathSoFar_);
+  pathSoFar_ = CGPathCreateMutable();
+  [self shapeLayer].path = nil;
   [self setNeedsDisplay];
 }
 
+- (void)dealloc {
+  CGPathRelease(pathSoFar_);
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.mm:276]: removing [super dealloc] is the only change.

**Review by chinmaygarde — APPROVED** [review]
cc @jason-simmons

**Comment by jmagman** [comment]
@hellohuanlin can you update the description and explain what it's relanding, and what is changed from the original one that was reverted?
Don't leave in paragraphs like "Replace this paragraph..."

**Review by jmagman — APPROVED** [review]

**Comment by hellohuanlin** [comment]
> @hellohuanlin can you update the description and explain what it's relanding, and what is changed from the original one that was reverted? Don't leave in paragraphs like "Replace this paragraph..."

Done!

**Pull request merged**